insights/client/crypto.py

```diff
@@ -0,0 +1,270 @@
+import errno
+import os.path
+import logging
+import shutil
+import tempfile
+import subprocess
+
+logger = logging.getLogger(__name__)
+
+
+class GPGCommandResult(object):
+    """Output of a GPGCommand.
+
+    :param ok: Result of an operation.
+    :type ok: bool
+    :param return_code: Return code of an operation.
+    :type return_code: int
+    :param stdout: Standard output of the command.
+    :type stdout: str
+    :param stderr: Standard error of the command.
+    :type stderr: str
+    :param _command: An optional reference to the GPGCommand object that created the result.
+    :type _command: GPGCommand | None
+    """
+    def __init__(self, ok, return_code, stdout, stderr, command):
+        self.ok = ok
+        self.return_code = return_code
+        self.stdout = stdout
+        self.stderr = stderr
+        self._command = command
+
+    def __str__(self):
+        return (
+            "<{cls} ok={ok} return_code={code} stdout={out} stderr={err}>"
+        ).format(
+            cls=self.__class__.__name__,
+            ok=self.ok, code=self.return_code,
+            out=self.stdout, err=self.stderr,
+        )
+
+
+class GPGCommand(object):
+    """GPG command run in a temporary environment.
+
+    :param command: The command to be executed.
+    :type command: list[str]
+    :param key: Path to the GPG public key to check against.
+    :type key: str
+    :param _home: Path to the temporary GPG home directory.
+    :type _home: str | None
+    :param _raw_command: The last invoked command.
+    :type _raw_command: list[str] | None
+    """
+    TEMPORARY_GPG_HOME_PARENT_DIRECTORY = "/var/lib/insights/"
+
+    def __init__(self, command, key):
+        self.command = command
+        self.key = key
+        self._home = None
+        self._raw_command = None
+
+    def __str__(self):
+        return "<{cls} _home={home} _raw_command={raw}>".format(
+            cls=self.__class__.__name__,
+            home=self._home,
+            raw=self._raw_command
+        )
+
+    def _setup(self):
+        """Prepare GPG environment.
+
+        :returns:
+            `True` if public GPG keys were imported into temporary environment,
+            `False` if there was an error.
+        :rtype: GPGCommandResult
+        """
+        # The /var/lib/insights/ directory is used instead of /tmp/ because
+        # GPG needs to have RW permissions in it, and existing SELinux rules only
+        # allow access here.
+        self._home = tempfile.mkdtemp(dir=type(self).TEMPORARY_GPG_HOME_PARENT_DIRECTORY)
+
+        logger.debug("setting up gpg temporary environment  in '{home}'".format(home=self._home))
+        result = self._run(["--import", self.key])  # type: GPGCommandResult
+        if not result.ok:
+            logger.debug("failed to import key '{key}': {result}".format(
```

Comment:
perhaps I'd bump this log message to warning, as it seems a severe enough failure